File: DP/Algorithms/Knapsack.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define rall(x) x.rbegin(), x.rend()
#define MOD 998244353
#define pi acos(-1)
#define gcd __gcd
#define lcm(a, b) ((a) * (b) / gcd(a, b))
#define precision(n) cout << fixed << setprecision(n)
#define debug(arr) cerr << #arr << " = "; for (auto &x : arr) cerr << x << " "; cerr << endl;
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

int main() {
    fastio;
    int n, W;
    cin >> n >> W;
    vector<int> wt(n), val(n);
    for (int i = 0; i < n; i++) {
        cin >> wt[i] >> val[i];
    }
    // dp[w] = máximo valor con capacidad w
    vector<ll> dp(W+1, 0);
    for (int i = 0; i < n; i++) {
        for (int w = W; w >= wt[i]; w--) {
            dp[w] = max(dp[w], dp[w - wt[i]] + val[i]);
        }
    }
    cout << dp[W] << "\n";
    return 0;
}

File: DP/Algorithms/Knapsack.py
import sys
input = sys.stdin.readline

# leer datos
n, W = map(int, input().split())
wt = [0]*n
val = [0]*n
for i in range(n):
    wt[i], val[i] = map(int, input().split())

# dp[w] = máximo valor con capacidad w
dp = [0] * (W+1)
for i in range(n):
    for w in range(W, wt[i]-1, -1):
        dp[w] = max(dp[w], dp[w - wt[i]] + val[i])

print(dp[W])
